pkg/datapath/linux/ipsec/ipsec_linux.go

```diff
@@ -700,13 +746,75 @@ func keyfileWatcher(ctx context.Context, watcher *fswatcher.Watcher, keyfilePath
 	}
 }
 
-func StartKeyfileWatcher(ctx context.Context, keyfilePath string, nodediscovery *nodediscovery.NodeDiscovery) error {
+func StartKeyfileWatcher(ctx context.Context, keyfilePath string, nodediscovery *nodediscovery.NodeDiscovery, nodeHandler datapath.NodeHandler) error {
 	watcher, err := fswatcher.New([]string{keyfilePath})
 	if err != nil {
 		return err
 	}
 
-	go keyfileWatcher(ctx, watcher, keyfilePath, nodediscovery)
+	go keyfileWatcher(ctx, watcher, keyfilePath, nodediscovery, nodeHandler)
 
 	return nil
 }
+
+func doReclaimStaleKeys() {
+	ipSecLock.Lock()
+	defer ipSecLock.Unlock()
+
+	xfrmStateList, err := netlink.XfrmStateList(0)
+	if err != nil {
+		log.WithField(logfields.SPI, ipSecCurrentKeySPI).
+			WithError(err).Warning("Failed to list XFRM states")
+		return
+	}
+
+	// First we need to find when the current key was added by looking at
+	// the AddTime of the IPSec state for the key.
+	//
+	// The creation of the IPSec state is asynchronous to loadIPSecKeys():
+	// - loadIPSecKeys() will only set ipSecCurrentKeySPI to the the SPI of
+	//   the new key
+	// - the actual IPSec state for the new key will be added when we
+	//   receive the k8s nodes from kube-apiserver and the local node from
+	//   our discovery
+	//
+	// Because of this, initialize currentKeyCreatedAt to the current time
+	// as we assume the current key is still being created unless we find
+	// its state.
+	currentKeyCreatedAt := time.Now()
+	for _, s := range xfrmStateList {
+		if s.Spi != int(ipSecCurrentKeySPI) {
+			continue
+		}
+
+		currentKeyCreatedAt = time.Unix(int64(s.Statistics.AddTime), 0)
+		break
+	}
+
+	// If the current key was added less then IPSecKeyDeleteDelay time ago,
+	// don't try to reclaim any stale state or policy
+	if time.Now().Sub(currentKeyCreatedAt) < linux_defaults.IPsecKeyDeleteDelay {
+		return
```

Comment:
I'm also curios about the situation where a new key is inserted to the k8s configmap, and a new node start up immediately. The new node will then only have the new key, while it can take a few minutes until all other nodes have the new key, since this solution will rely on the kubelet updating the configmap on disk.

In such a situation, will a new node be blocked from scheduling pod networked traffic, or will it be able to schedule pods successfully, but they won't have pod network connectivity the first few minutes.. 🤔  Have't had time to test it yet.